src/exa.c

```diff
@@ -397,60 +715,914 @@ static void TegraEXACopy(PixmapPtr pDstPixmap, int srcX, int srcY, int dstX,
+
+    TegraPushVtxAttr(dst_left,  dst_bottom,  true);
+    TegraPushVtxAttr(src_left,  src_bottom,  push_src);
+    TegraPushVtxAttr(mask_left, mask_bottom, push_mask);
+
+    TegraPushVtxAttr(dst_left,  dst_top,  true);
+    TegraPushVtxAttr(src_left,  src_top,  push_src);
+    TegraPushVtxAttr(mask_left, mask_top, push_mask);
+
+    TegraPushVtxAttr(dst_right,  dst_top,  true);
+    TegraPushVtxAttr(src_right,  src_top,  push_src);
+    TegraPushVtxAttr(mask_right, mask_top, push_mask);
+
+
+    TegraPushVtxAttr(dst_right,  dst_top,  true);
+    TegraPushVtxAttr(src_right,  src_top,  push_src);
+    TegraPushVtxAttr(mask_right, mask_top, push_mask);
+
+    TegraPushVtxAttr(dst_right,  dst_bottom,  true);
+    TegraPushVtxAttr(src_right,  src_bottom,  push_src);
+    TegraPushVtxAttr(mask_right, mask_bottom, push_mask);
+
+    TegraPushVtxAttr(dst_left,  dst_bottom,  true);
+    TegraPushVtxAttr(src_left,  src_bottom,  push_src);
+    TegraPushVtxAttr(mask_left, mask_bottom, push_mask);
+
+    scratch.vtx_cnt += 6;
+}
+
+static void TegraEXADoneComposite(PixmapPtr pDst)
+{
+    ScrnInfoPtr pScrn = xf86ScreenToScrn(pDst->drawable.pScreen);
+    TegraEXAPtr tegra = TegraPTR(pScrn)->exa;
+    struct tegra_fence *fence = NULL;
+    TegraPixmapPtr priv;
+
+    if (scratch.solid2D)
+        return TegraEXADoneSolid2D(pDst);
+
+    TegraEXACompositeDraw(&tegra->cmds);
+
+    if (scratch.ops && tegra->cmds.status == TEGRADRM_STREAM_CONSTRUCT) {
+        if (scratch.pSrc) {
+            priv = exaGetPixmapDriverPrivate(scratch.pSrc);
+
+            if (priv->fence && priv->fence->gr2d) {
+                TegraEXAWaitFence(priv->fence);
+
+                tegra_stream_put_fence(priv->fence);
+                priv->fence = NULL;
+            }
+        }
+
+        if (scratch.pMask) {
+            priv = exaGetPixmapDriverPrivate(scratch.pMask);
+
+            if (priv->fence && priv->fence->gr2d) {
+                TegraEXAWaitFence(priv->fence);
+
+                tegra_stream_put_fence(priv->fence);
+                priv->fence = NULL;
+            }
+        }
+
+        priv = exaGetPixmapDriverPrivate(pDst);
+        if (priv->fence) {
+            if (priv->fence->gr2d)
+                TegraEXAWaitFence(priv->fence);
+
+            tegra_stream_put_fence(priv->fence);
+            priv->fence = NULL;
+        }
+
+        tegra_stream_end(&tegra->cmds);
+
+        /*
+         * XXX: Glitches may occur due to lack of support for waitchecks
+         *      by kernel driver, they are required for 3D engine to complete
+         *      data prefetching before starting to render. Alternative would
+         *      be to flush the job, but that impacts performance very
+         *      significantly and just happens to minimize the issue, so we
+         *      choose glitches to low performance. Mostly fonts rendering is
+         *      affected.
+         *
+         *      See TegraGR3D_DrawPrimitives() in gr3d.c
+         */
```

Comment:
Are we sure that we should merge this without this problem sorted out first?

Reply:
It is not so bad. Sometimes when text is drawn, some words getting cropped; but you can just select text or move window around to cause the re-drawing, not a big deal. Should be fine, we can add a Xorg configuration option for the driver that disables compositing.